src/protocols/LoRaWAN/LoRaWAN.cpp

```diff
@@ -749,23 +942,55 @@ int16_t LoRaWANNode::downlink(uint8_t* data, size_t* len) {
       remLen -= processedLen;
       foptsPtr += processedLen;
     }
+
+    // if FOptsLen for the next uplink is larger than can be piggybacked onto an uplink, send separate uplink
+    if(this->commandsUp.len > 15) {
```

Comment:
I'm not sure it's a good idea to automatically send an uplink in this case - if we do, the device should then open the RX windows, right?

Reply:
I agree that this may not be the best place to send the uplink - it's probably best to complete the downlink first. But I couldn't really come up with an elegant way to separate the uplink from the downlink. We would maybe need to 'signal' the user to trigger an uplink from the main code, but if that doesn't happen, the MAC commands will only pile up worse and worse.

Reply:
I suspect that the listen for downlink is a MUST. As eluded to above, not returning to the user until commands are flushed out isn't unusual as functionality, even if rarely experienced. I think that's how LMIC does it.

Reply:
The main thing that I'm worried about here is that we end up taking control from the user for a long time. I don't know how likely is the situation of user uplinks not being able to keep up with MAC commands.

The way I see it there are three options now:

1. Don't do the uplink automatically at all and let the user know that the queue is full. We could then have a method e.g. `flushMacCommands` that the user can call to do this MAC-only uplink/downlink.
2. Keep the uplink here and signal to user (e.g. by return code) that he needs to call downlink again. I don't like that too much since it seems to kind of break the normal flow from the users perspective - first they call uplink, then downlink and then they need to call downlink again for a reason that is out of their control.
3. Add the downlik here. This blocks for the longest amount of time, but in the current state it is likely the easiest and the cleanest solution.

Future solution would have to include interrupt-driven processing of the commands so that all of this really does happen in the background (as the standard calls for).

Reply:
> I don't know how likely is the situation of user uplinks not being able to keep up with MAC commands.

Very unlikely - it's not a use case for LoRaWAN for people to try to keep up with the legal duty cycle and it's an extremely rare event that you end up with a pile of MAC commands moving up & down - you can trigger your own mess with a LinkCheckReq, a TimerReq, an ADR update, but a device getting a downlink after the initial foray at JoinAccept & ADR adjustments is rare - if everything is working and the RF is stable so no ADR changes, then a device can go months without a downlink.

> Future solution would have to include interrupt-driven processing of the commands so that all of this really does happen in the background 

But it will still block the user's use of the radio.

The processing time should be very short, it's giving the user back some cycles whilst waiting for Rx1 that people ask for but then never really make good use of. You can't run a real time system doing something really important if you need to hit the Rx1 & 2 windows, but again, with some evaluation, if the system is serviced before initiating committing to the radio, the system gets some time during the delay, if the Rx1 window gets trashed by an interrupt it's no worse than RF interference stopping reception. 

> (as the standard calls for).

Which standard?

MAC commands coming down can be processed as a low priority task, they just have to be enacted before the next uplink.

Reply:
Thanks @HeadBoffin for the inputs - in this case I think going with option 3 from my previous post is the best option now. @StevenCellist this is the major thing blocking this PR.

Also, @HeadBoffin 

> Which standard?

I meant the LoRaWAN specification, which explicitly states the following - I understood that as meaning that the user/application should not be required to do anything (maybe apart from the initial join action) in order for MAC commands to work.

![Screenshot_68](https://github.com/jgromes/RadioLib/assets/3407192/3573a20a-a060-493a-b9eb-7483f5c29b7b)

Reply:
Sorry, now I'm the one short on time :)
I'll get the PR fixed in the weekend, I expect! And after that work on the other stuff mentioned in the issue